`unobserveproperty` operation - closes #56

benfrancis opened the pull request:
Closes #56.

This PR defines request, response and notification messages for the unobserveproperty operation, as discussed in in #42 and #56.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/benfrancis/web-thing-protocol/pull/86.html" title="Last updated on Sep 17, 2025, 4:43 PM UTC (41ff84d)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/web-thing-protocol/86/598f68b...benfrancis:41ff84d.html" title="Last updated on Sep 17, 2025, 4:43 PM UTC (41ff84d)">Diff</a>

benfrancis commented:
I think, having specified this, that my instinct is that `observeproperty` and `observeallproperties` subscriptions should be treated separately, so that after an `unobserveproperty` a Consumer will still receive notifications for all properties relating to an `observeallproperties` operation, until an `unobserveallproperties` operation occurs. The Consumer will know which subscription a notification relates to due to the value of the `operation` member of a notification message.

But does this mean duplicate notifications should be sent for overlapping observations?

hspaay commented:
> my instinct is that observeproperty and observeallproperties subscriptions should be treated separately
> ...
> But does this mean duplicate notifications should be sent for overlapping observations?
> See also: https://github.com/w3c/web-thing-protocol/issues/41 where we reached a different conclusion.

What has the lowest risk of a consumer making a mistake. What is the easier mental model? 
As an implementor I am looking for the request that (un)subscribes the properties I'm interested in. This can be individual subscription and bulk subscriptions (also think of observemultipleproperties).
My mental model tracks the subscriptions of the properties, not which operation is used. 

So in short, I think that observeall(multiple) should not be tracked separately from observeproperty.

hspaay reviewed (changes requested):
I request to add the following description of behavior:

1. Similar to observeproperty, unobserveproperty is idempotent. Invoking unobserveproperty multiple times provides the same result as the first time. It is not an error to invoke it multiple times as the requested result is achieved.  A property has only a single observation registration per connection. 

2. If unobserveproperty is invoked after observeallproperties or observemultipleproperties then only the registration of the mentioned property is removed. 

3. If unobserveproperty is invoked after bulk operations such as unobserveallproperties or unobservemultipleproperties then the result is success but no registration has changed since the property was already unobserved.

4. If a consumer has established multiple websocket connections then only the observation registration of the connection through which the operation is sent is affected. Observation of properties on any other connections are not affected.

benfrancis commented:
@hspaay wrote:
> Similar to observeproperty, unobserveproperty is idempotent. Invoking unobserveproperty multiple times provides the same result as the first time. It is not an error to invoke it multiple times as the requested result is achieved

I agree, except that I think if a Consumer sends an `unobserveproperty` request for a Property which has no active observation subscriptions, it should receive an error (e.g. 404) because the observation subscription does not exist and therefore the operation is not possible. Reporting success may give the impression that the subscription has only just been cancelled and that any property changes up until this point would have resulted in a notification.

> If unobserveproperty is invoked after bulk operations such as unobserveallproperties or unobservemultipleproperties then the result is success but no registration has changed since the property was already unobserved.

As above, I think the Consumer should receive an error under these circumstances, since the operation is not possible.

hspaay commented:
The clarification looks good. 

One more thought came to mind that is worth mentioning: If the connection breaks then what happens to the subscriptions? I prefer to automatically remove all subscriptions made through that connection.  This is the equivalent to MQTT's retain=false, setting and much simpler to implement and manage.

Wrt to the error behavior, technically you are correct, if the subscription is seen as a resource and it isn't there then it isn't found. So I get your point. I also think this is not the best behavior:

1.  The desired effect - not having a subscription - is achieved successfully. From an application point of view there is no error.
2.  It is a bit of a rabbit hole. Following that line of reasoning then unobserveallproperties should fail if at least one property has no subscription. This would defeat the purpose of this bulk operation so in this case it is okay?. This is inconsistent.
3.  What is the consumer supposed to do when an error is returned by unobserveproperty? You're implying that it is a bug in the consumer if unobserve(all)properties is invoked twice and that the consumer should fix its implementation.

If an error is returned then point 3 should be stated clearly. Consumers MUST only invoke unobserveproperty once per property after observe(all)properties, which is not helping anyone.

To me this is an edge case that can easily be handled gracefully without putting a burden on the consumer, assuming there is no use-case where an error is needed.

benfrancis commented:
@hspaay said:
> One more thought came to mind that is worth mentioning: If the connection breaks then what happens to the subscriptions? I prefer to automatically remove all subscriptions made through that connection. This is the equivalent to MQTT's retain=false, setting and much simpler to implement and manage.

I think I probably agree, but let's discuss that separately in #69.

> Wrt to the error behavior, technically you are correct, if the subscription is seen as a resource and it isn't there then it isn't found. So I get your point. I also think this is not the best behavior

OK, based on feedback from you and Ege I have removed the error response and added an assertion to clarify that the Thing should send a success response even if there are no active observation subscriptions for that Property.

@hspaay @egekorkan Are you happy with this PR now?

hspaay reviewed (approved):
Looks good, thank you.

benfrancis commented:
@egekorkan Are you OK with this PR now?

benfrancis commented:
Thank you for the reviews!
